mitiq/cdr/data_regression.py

```diff
@@ -0,0 +1,294 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+""" Functions for using near-Clifford training data for error mitigation"""
+from typing import List, Union, Callable, Dict
+
+import numpy as np
+
+from cirq.circuits import Circuit
+
+from scipy.optimize import curve_fit
+
+
+def scale_noise_in_circuits(
+    circuits: Union[Circuit, List[Circuit]],
+    scale_noise: Callable[[Circuit, float], Circuit],
+    scale_factor: Union[float, List],
+) -> List[Circuit]:
+    """Returns a list consisting of the original circuit followed by the circuit
+    folded by the specified method and the scale factors.
+
+    Args:
+        circuits: list of training circuits.
+        scale_factor: factor by which to scale the circuits.
+        scale_noise: method to use to fold the circuits.
+    """
+    if isinstance(circuits, Circuit):
+        circuits = [circuits]
+    folded_circuits = []
+    if isinstance(scale_factor, float) or isinstance(scale_factor, int):
+        scale_factor = [scale_factor]
+    for scale_factor_ in scale_factor:
+        for circuit in circuits:
+            folded_circuits.append(scale_noise(circuit, scale_factor_))
+    return circuits + folded_circuits
+
+
+def execute_training_circuits(
+    all_training_circuits: List[Circuit],
+    executor: Callable[[Circuit], Dict],
+    simulator: Callable[[Circuit], Union[Dict, np.ndarray]],
+    noise_levels: int = 1,
+) -> (List[Dict], List[Dict]):
+    """Returns two list of dictionaries of different sizes. One list of
+    dictionaries consisting of the real counts at various noise levels for the
+    training circuits and the other contains the exact counts.
+
+    Args:
+        all_training_circuits: list of all training circuits at all noise
+                               levels.
+        executor: user defined executor function.
+        simulator: user defined simulator function either returns counts or
+                   a statevector.
+        noise_levels: number of noise levels to be used in the regression.
+    """
+    number_of_training_circuits = int(
+        len(all_training_circuits) / noise_levels
+    )
```

Comment:
This is extremely confusing. If I understand right, the logic is:

- Run all circuits with `executor`.
- Run circuits with scale factor 1 with `simulator`.

If so, just get rid of this helper function altogether. You probably want to store the circuits in a way that's amenable to what you want to do. How about a 2d structure with dimensions # of training circuits x # of scale factors? Then you just pass in the first column to `simulator` and the all circuits to `executor`. 

```python
circuits =  2d data structure described above
executor_data = [executor(c) for c in circuits.flatten()]
simulator_data = [simulator(c) for c in 1st column of circuits]
```

It's much shorter and much more readable.

This should take into account batching as well, but let's go one step at a time.

Reply:
for this 2d structure what would you recommend. Is a lists of lists okay?